### ISSUE TITLE
Stack overflow on EOF*

### ISSUE THREAD

**Issue by KvanTTT** [antlr/antlr4#3359]
Consider the following grammar:

```g4
lexer grammar StackoverflowOnClosure;
ClosureEof: EOF*;
```

It fails with Stackoverflow on empty input:

```
at org.antlr.v4.runtime.misc.Array2DHashSet.getOrAdd(Array2DHashSet.java:59)
	at org.antlr.v4.runtime.atn.ATNConfigSet.add(ATNConfigSet.java:146)
	at org.antlr.v4.runtime.atn.ATNConfigSet.add(ATNConfigSet.java:122)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:446)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
```

And it's freezing on any not empty input.

The samples is extracted from https://github.com/antlr/antlr4/issues/1943#issuecomment-974566196

**Comment by KvanTTT** [comment]
It seems like the bug is similar to [the bug with `EPSILON_CLOSURE`](https://github.com/antlr/antlr4/issues/2860) that I've already fixed. But here we have `EOF_CLOSRE` instead of `EPSILON_CLOSURE`.

So, I fixed it in a similar way.

**Comment by parrt** [comment]
Hmm..gotta be careful here. EOF is an actual condition/"char" that we sometimes wanna match.  maybe not in `*` loop but certainly in a set

**Comment by KvanTTT** [comment]
It works fine with set. There is no error for such case:

```g4
EofInSet: 'y' ('z' | EOF);
```

But the following code where EOF closure actually exists:

```g4
prog : stat EOF;
stat : 'x' ('y' | EOF)*?;
```

Should be rewritten in the following way:

```g4
prog : stat EOF;
stat : 'x' 'y'* EOF?;
```

**Comment by parrt** [comment]
Should this be merged before or after your #3349 ?

**Comment by KvanTTT** [comment]
The PR fixes this bug :)
